src/abstract_domains.rs

```diff
@@ -32,9 +112,93 @@ pub enum ExpressionDomain {
     /// that always panics.
     Bottom,
 
-    /// An expression that is a compile time constant value, such as a numeric literal or a
-    /// function.
+    /// An expression that is a compile time constant value, such as a numeric literal or a function.
     CompileTimeConstant(ConstantValue),
+
+    /// An expression that is either if_true or if_false, depending on the value of join_condition.
+    JoinedExpression {
```

Comment:
The name JoinExpression and JoinCondition is a little bit weird to me. You have an abstract domain for a condition, and based on that domain you may be able to assume if_true or if_false, but why does it matter that you get it from a join? As in, why isn't this just called Condition { condition: ... }?

It feels like maybe two things are being mixed here. In the AbstractDomain trait you have join with a condition field. I guess that makes some sense, if a lot of joins have extra information about when a join should actually be done or not. But here it sounds like it's more about if/while/etc statements. Or is it meant more generically here than branching checks?